### ISSUE TITLE
reopen issue #10117 - IIS app pool recycle throws 503 errors

### ISSUE THREAD

**Issue by alex-jitbit** [dotnet/aspnetcore#41340]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

IIS app pool throws 503 errors during recycles. This is a known issue with ANCM module that has been reported previously in #10117 - which has 33 likes and a 3-year discussion, it was never fixed, but it was automatically closed by a bot "as a clean-up due to lack of discussion".

P.S. This is not a deployment problem. There are many scenarios when IIS app pool is being recycled outside of our control (adding/removing SSL certificates, changing IP addresses to listen to, etc... basically, touching any IIS setting causes a recycle - and 503 errors are unacceptable for high-availability scenarios).

.NET Framework was free of this bug.

### Expected Behavior

No errors during recycles.

### Steps To Reproduce

see the issue linked #10117

### .NET Version

5.0, 6.0 7.0 8.0

**Comment by Tratcher** [comment]
https://github.com/dotnet/aspnetcore/issues/10117#issuecomment-498865444

> Using separate app pools and a load balancer is our recommended approach for high-availability as it allows you full flexibility over deployment process and the ability to easily revert versions.

Trying to achieve high-availability with a single instance is not recommended.

**Comment by alex-jitbit** [comment]
@Tratcher like I indicated above, this is not about _deployments_. There are a lot of scenarios when IIS recycles the pool (see above)

**Comment by Tratcher** [comment]
Deployments are just one example that disrupt availability. A single instance is not advised for high-availability for many reasons.

**Comment by benjamin-stern** [comment]
@Tratcher Even having more than a single instance this would still strongly affect a service, as all the requests going to the server that's recycling would be returned the 503 error.

**Comment by c0shea** [comment]
We have two instances running behind a load balancer and have still experienced this issue intermittently. When the app pool inevitably recycles (due to deployment, config change, etc), it starts returning 503 instead of queuing up the requests.

The load balancer doesn't immediately treat the 503s as the server being down and take it out of the rotation. Instead, it uses a polling mechanism that calls an endpoint (i.e. /status) on each instance and checks for a successful response. While that status endpoint is monitored frequently enough, there is obviously plenty of time where a bunch of requests will fail with 503 while the recycle is happening. We can't have the load balancer take the instance out of the rotation if it sees 503s being returned because (1) it's not an available option in NetScaler and (2) if both happen to recycle at the same time, both servers would be taken out of the rotation and the service would be completely down until manual intervention told the load balancer that the requests aren't failing anymore.

**Comment by RomBrz** [comment]
In this scenario (using two or more instances behind a load balancer), since you're using a "/status" to check availability, i suggest that you make some routines during the deployment or maitenence to, before start doing anything, force the "/status" to throw an "unhealthy" status, so the load balance could remove the node from the balancer and then make the changes.

About the issue itself, the IIS default behavior on a recycle is to first start a new application pool, route the new requests to the new application pool, wait the default set time to the current requests ends, an than close/finish the current application pool, keeping only the new as the application pool.

Recycling an application pool, could be an ASPNET Core "expected behavior", but looking at IIS, throwing 503 during one recycle isn't a "normal behavior".

**Comment by c0shea** [comment]
> 

The problem is that ASP.NET Core doesn't use the overlapping recycle behavior that .NET Framework did. While the old worker process is being shutdown (especially if there were a lot of inflight requests being handled by it), the new process isn't yet started and those requests in the middle get the 503.

**Comment by alex-jitbit** [comment]
TIL that StackOverflow also runs ASP.NET Core under IIS

![image](https://user-images.githubusercontent.com/33555768/174447160-29804394-f8b9-4093-98ca-4838edf88bb4.png)

**Comment by luizfbicalho** [comment]
Is there any way to minimize this problem, or ate least to detect what is causing it?

**Comment by peter-bertok** [comment]
> Deployments are just one example that disrupt availability. A single instance is not advised for high-availability for many reasons.

High availability and throwing 503s from otherwise "perfectly fine servers" are separate concerns.

Most load balancers do not hide HTTP errors! If the IIS process responds to a HTTP request with 503, then that's what the user will see. In particular, none of the Azure load balancer offerings hide errors from the users. They pass them on faithfully.

This behavior is triggered by many actions, not all of which are resolved via hosting on multiple server instances. Scheduled recycling, for example, has been mentioned by many people as a common trigger.

**Comment by luizfbicalho** [comment]
My problem seems to be on IIS recycle when takes too long to recycle, it's not related to cpu neither to memory, I asked the IT infrastructure to add more performance counters to grafana but they didn't add yet, 

I'm inclined to think that the problem is with Network connections, what do you think that I should monitor in grafana?

**Comment by alex-jitbit** [comment]
TIL that `fuget.org` also uses IIS + ASP.NET Core, just caught them during a recycle.
![image](https://user-images.githubusercontent.com/33555768/199351561-654fbdff-5c0a-42b2-9dc1-2648e7831e16.png)

How many more examples do we need?

**Comment by AndreasJilvero** [comment]
In my experience, this happens also when simply setting the physical path of a website. The response is somewhat difference though - a recycle renders the text "The service is unavailable" and setting the physical path just gives an empty 503 result.

https://stackoverflow.com/questions/74326315/iis-setting-physical-path-gives-503-status-for-a-few-seconds

**Comment by luizfbicalho** [comment]
In my case it's not for seconds, after the first 503 only iisreset solves the problem

**Comment by kadamgreene** [comment]
We are also experiencing this issue.  Is there any expectation of a fix?  This is going to keep us from moving forward with .NET migration / any new work in .NET 6+.  It's not the deployments, blue/green can handle that, it's the "unexpected" recycles in the run of a day that are the issue.

**Comment by cun-dp** [comment]
We are also seeing this with our nightly apppool recycles in the production environment, and with all our aspnetcore microservices.

Example of IIS Logs (anonymized): 
2023-04-23 22:47:03 W3SVC5 HOSTNAME1 [IP censored] GET /app/health 443 - HTTP/1.1 - - my.fqdn.example.com **503 0 1255** 202 130 15

HTTP 503 with win32 substatus 1255. According to https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes--1000-1299- code 1255 matches: 1255 (0x4E7) ERROR_SERVER_SHUTDOWN_IN_PROGRESS: The server machine is shutting down.

**This is a critical issue for us, since we are running a 24/7 plattform.**

**Comment by Tratcher** [comment]
@cun-dp that makes sense, the application has stopped serving traffic and can't re-start until the current process exits.

**Comment by luizfbicalho** [comment]
> @cun-dp that makes sense, the application has stopped serving traffic and can't re-start until the current process exits.

but the correct approach wasn't to start a new process, redirect the new connections to that process and let the old process die in peace as long as it takes?


Is there a way for me to see what's blocking the old process to die?

Is there a way to force all the threads in the old process do exit?

Is there any workaround to help with this problem?

**Comment by gcbenjamin** [comment]
> And to make this abundantly clear, because it has been misinterpreted in #10117 a lot: This is **not** about deployments of applications. This bug is triggered by **only recycling** an app pool running an **existing, unchanged, aspnetcore application**.
> 
> The behaviour can be reproduced by just running `Restart-WebAppPool MyAppPool` in PowerShell while continuously sending request to the app running in "MyAppPool".

We see the exact same thing and easy to reproduce as explained by @cun-dp , just recycle while under load and requests will fail with 503's. This behaviour is not seen in any of our framework api's, only .net core. I've tried all different combinations of IIS/App pool settings and nothing has worked. Rant he default weather app created in Visual Studio new project, hit the get endpoint continuously using K6 and always get hit with 503's when recycling under load. The API just returns text, no db's or anything complex.

**Comment by estebanorellana** [comment]
In our work, the exact same thing happens to us when we install the .NET 6 application on a server with IIS, after the first 503 error it does not come out without an iisreset.

but if we raise this site with kestrel we have no problem.

**Comment by alex-jitbit** [comment]
Looking at the [ANCM commit history](https://github.com/dotnet/aspnetcore/commits/main/src/Servers/IIS/AspNetCoreModuleV2/AspNetCore) I see it had **_2 commits in 1 year_** I'm thinking MS priorities are elsewhere

(or the C++ guru who wrote it has left the company and now everyone's just afraid to touch it)

**Comment by luizfbicalho** [comment]
I saw that there is a @BrennanConroy that is commiting code, It would be great if we could get a better error message that is preventing the shutdown of the aspnet core app.

**Comment by petersladek** [comment]
This clearly is deterioration vs ASP.NET behavior in IIS and it catches most people by surprise in production (doesn't matter if one is using dedicated load balancer or not as typically there will be http 503 responses passed through LB to client before LB removes the traffic to unhealthy node). Can we at least get the documentation updated to state this difference in behavior of ASP.NET Core apps in IIS vs ASP.NET Framework apps in IIS? And maybe suggesting that when running ASP.NET Core apps in IIS one should consider disabling periodic recycling of app pool (as **default** is to recycle every 1740 minutes) to avoid http 503 errors.

**Comment by divil5000** [comment]
We are running into this issue in production too. Our service is constantly serving requests, and changing the physical path (to update the binaries) leads to many seconds of 503 responses before it starts serving successfully again. I cannot believe Microsoft are ignoring this issue.

**Comment by luizfbicalho** [comment]
> We are running into this issue in production too. Our service is constantly serving requests, and changing the physical path (to update the binaries) leads to many seconds of 503 responses before it starts serving successfully again. I cannot believe Microsoft are ignoring this issue.

My problem is worse, when I receive 503 que app pool stop and doesn't restart

**Comment by TRMack** [comment]
We are in the same boat; we run behind a load balancer but a scheduled app pool recycle on an instance -- it recycles every 29 hours -- returned 503 to a customer which disrupted an automated process of theirs. I always trusted IIS to handle the recycle gracefully and am frustrated to see that that is no longer the case.

**Comment by cun-dp** [comment]
[...]
> While this doesn't fix the problem, it has helped, in the app pools advanced settings, setting Disable Overlapped Recycle to `TRUE` I get around 90% less 503's (in my tests from 60 to 7).

I could not reproduce this mitigation. 

In my tests, the differences were well within the variance I would expect.
I did 7 testruns, each once with OR ON and OR OFF (with IIS Resets inbetween for the last 2 runs, which didnt make a difference either). 

My test: Send GET requests to the application as fast as possible. Recycle the apppool 10 times, waiting 5 seconds between each recycle.
Results (# of HTTP503 responses):
Overlapped Recycle ON: 591, 691, 741, 712, 736, 631
Overlapped Recycle OFF:  561, 677, 748, 696, 621, 598

**Comment by luizfbicalho** [comment]
I just want an official microsoft statement, if you use this, do that, change to docker, or something else

Did anyone get any document about that? I need to show something to my customer

**Comment by MV10** [comment]
We run _hundreds of thousands_ of IIS servers, and some of our load balanced farms span 120+ servers in multiple global data centers, and this is becoming a very serious problem as apps (finally) migrate away from .NET Framework. 

As **many** others have stated clearly, this is almost entirely unrelated to deployment. With a couple tens of thousands of applications in production usage, our operations people do several hundred overlapped recycles of unhealthy applications every day, and most of our applications are on staggered nightly recycle schedules.

We haven't had much luck getting our very-expensive Microsoft DSEs to help prioritize anything in the .NET from a dev/issue perspective (I'll spare you my rants about Agile perma-backlogged deferral), but I'll be running this one up the chain with management again. There's always a first time, I guess.

Also, the `area-networking` tag is incorrect in my opinion. This is runtime process behavior. Or maybe something to do with IIS interaction, if such a tag exists.

Whatever the tag, this is extremely critical for real-world production business usage.

**Comment by alex-jitbit** [comment]
# This problem has become way worse in .NET 8!

Before net8 - the becomes sluggish and throws 5-10 errors 503.

After net8 - it seems like all requests throw 503, until the recycled process spawns up.

Not only you haven't fixed this, you even made it worse!

**Comment by MV10** [comment]
We were told this is "by design" and won't be fixed.

Insane.

**Comment by luizfbicalho** [comment]
> We were told this is "by design" and won't be fixed.
> 
> Insane.

Where they told you this? If you find this for me is better to convince my client that this won't work anymore

**Comment by MV10** [comment]
@luizfbicalho 
Our DSEs (Dedicated Support Engineers) quoting whatever internal sources they have.

**Comment by luizfbicalho** [comment]
> @luizfbicalho Our DSEs (Dedicated Support Engineers) quoting whatever internal sources they have.

Thanks, but i needed some oficial answer

**Comment by MV10** [comment]
DSEs are MS employees, to be clear -- but I know you can't just quote "some random guy on the Internet" (me).

**Comment by luizfbicalho** [comment]
yeah, would be nice if microsoft had an official statement abou this, and this client could use linux because it doesn't use anything from iis

**Comment by Hieu-Nguyen-1** [comment]
> # This problem has become way worse in .NET 8!
> Before net8 (we were on net6) - the app becomes sluggish and throws 5-10 errors 503.
> 
> After moving from net6 to net8 - it seems like all requests throw 503, until the recycled process spawns up.
> 
> Not only you haven't fixed this, you even made it worse

I think that it will be fixed in .NET 8! But.....!!!! Please fix it @Microsoft

**Comment by alex-jitbit** [comment]
OP here.

Welp.

It's been more than 4 years of Microsoft ignoring this issue (if you count the original one). Even though the 503 errors can be seen on major .NET-powered websites like StackOverflow (see example screenshots above in this issue). StackOverflow, by the way, is proudly showcased as Microsoft's _number one_ customer on their ".NET Customers Showcase" page https://dotnet.microsoft.com/en-us/platform/customers/aspnet

I still love .NET and C# too much to abandon it, so I will continue to use it.

However as of this weekend we have moved all our production servers away from Windows to Linux. We're not paying for Windows Server licenses any more. If Microsoft can't make their own two products work together - well, you just lost a loyal paying customer.

💔

**Comment by luizfbicalho** [comment]
> OP here.
> 
> Welp.
> 
> It's been more than 4 years of Microsoft ignoring this issue (if you count the original one). Even though the 503 errors can be seen on major .NET-powered websites like StackOverflow (see example screenshots above in this issue). StackOverflow, by the way, is proudly showcased as Microsoft's _number one_ customer on their ".NET Customers Showcase" page https://dotnet.microsoft.com/en-us/platform/customers/aspnet
> 
> I still love .NET and C# too much to abandon it, so I will continue to use it.
> 
> However as of this weekend we have moved all our production servers away from Windows to Linux. We're not paying for Windows Server licenses any more. If Microsoft can't make their own two products work together - well, you just lost a loyal paying customer.
> 
> 💔

Can you detail more what you are going to use? what linux? what webserver? configuration that you changed from the default

**Comment by alex-jitbit** [comment]
> Can you detail more what you are going to use? what linux? what webserver? configuration that you changed from the default

@luizfbicalho Our config is very basic and follows the official MS [docs](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-nginx?view=aspnetcore-8.0&tabs=linux-ubuntu)

1. Ubuntu 22.04 with Nginx as a reverse proxy in front of the app (that runs on port 5002)
2. A `systemd` "service" described by one file `/etc/systemd/system/MyApp.service`, which simply runs `dotnet MyApplicaion.dll` and sets some basic settings.

That's it. Very simple. Just make sure you call `UseForwardedHeaders` in your app to make "reverse-proxy-firendly".

Whenever nginx configuration needs to change (a new IP to listen to, a new SSL certificate installed, a new hostname to listen to etc etc) you just change nginx config files and run `sudo service nginx reload` with zero downtime.

P.S. We've actually implemented _two_ systemd-services for blue-green deployment that run on different ports, but that's another story.

**Comment by BrennanConroy** [comment]
Hey folks, I looked into this and made some improvements, at least with local testing. I'd love for people on this thread who are seeing 503s during recycles to voluntarily test a change.

At the bottom of this comment is a dll with some changes to try and improve 503s during overlapped recycles. The dll has been signed by Microsoft, you can verify this by running `signtool verify /v /pa aspnetcorev2.dll`. Signtool is located in `C:\Program Files (x86)\Windows Kits\10\bin\10.0.22621.0\x64\signtool.exe` (note: 10.0.xxxxx.0 version may vary depending on your machine). The SHA256 for the dll is `BECDEC71EE95A2F117A75A5F2BF961EED3EEC1A641D8A4573EC53AEE25696AE2`

For those of you who would prefer building your own dll, or want to preview the changes, you can check out https://github.com/dotnet/aspnetcore/pull/52807 and build via build.cmd in /src/Servers/IIS/.

**To use the dll below** (assuming x64), replace the dll in `C:\Program Files\IIS\Asp.Net Core Module\V2\` (keep the original around so you can switch back to the supported dll later).

There is also an **optional config** you can change to influence how slow (or fast) shutdown occurs if you're running in a slower environment and need more time for shutdown to avoid the 503s. The option is `shutdownDelay` which can be set in web.config. The default is 1000ms (1 second).
```xml
<aspNetCore processPath="dotnet" arguments="myapp.dll" stdoutLogEnabled="false" stdoutLogFile=".logsstdout">
      <handlerSettings>
        <!-- Milliseconds to delay shutdown by, this doesn't mean incoming requests will be delayed by this amount, but the old app instance will shutdown after this timeout occurs -->
        <handlerSetting name="shutdownDelay" value="5000" />
      </handlerSettings>
    </aspNetCore>
```

Note: **Use of this dll is purely experimental and at your own risk**, if problems like crashes occur please let us know, so we can improve the final code.
[aspnetcorev2.zip](https://github.com/dotnet/aspnetcore/files/13680906/aspnetcorev2.zip) (github doesn't allow uploading .dll so the dll is in a zip file)

**Comment by luizfbicalho** [comment]
I think that my problem is related to the [https://github.com/HangfireIO/Hangfire/issues/1345 ](https://github.com/HangfireIO/Hangfire/issues/1345 )

**Comment by JuergenAuer** [comment]
There is a crazy workaround.

I'm running a web application, 17 years, .NET-Framework, code updates or nightly recycling - never a problem. After a restart - the next call is slower, but it worked.

Now switched to NET.8, only the test system. Checking one page with a command line tool, recycle - 23 http status 503. If a user posts data, he has to do it again, that's not a solution. Slow is ok, but a 503 is deadly.

Found this topic.

Buggy configuration:

Root: E:\net-www + web.config, E:\net-www\bin with a lot of dll, web.config:

```
      <aspNetCore processPath="dotnet" arguments=".\bin\this_webserver.dll" stdoutLogEnabled="true" 
stdoutLogFile=".\logs\stdout" hostingModel="InProcess">
	    <handlerSettings>
	        <handlerSetting name="EnableShadowCopy" value="true" />
	        <handlerSetting name="shadowCopyDirectory" value="E:/temp/ShadowCopyDirectory/" />
	    </handlerSettings>
	</aspNetCore>
```


Now:

1. Change the hostingModel to OutOfProcess. Result: Started command line fetch, recycle - no 503. First call is slow, then it's ok, no significant difference to InProcess.

2. The dll are blocked, updates are not possible. No shadow copy. But:

3. Created two subdirectories E:\net-www\bin1, E:\net-www\bin2, copy all files from E:\net-www\bin into bin1/bin2.

4. If required, switch arguments=".\bin\this_webserver.dll"  to arguments=".\bin1\this_webserver.dll" or arguments=".\bin2\this_webserver.dll". Or save these files somewhere and copy the file, so the process is recycled.

Conclusion: OutOfProcess waits instead of throwing a 503, bin1/bin2 is like a green-blue-configuration. IIS selects a new port (random).

**Comment by ckobelski** [comment]
@BrennanConroy we tried out the updated DLL in one of our testing environments and got really promising results. We set up a a few constant streams of requests to one of our .Net 6 services hosted in IIS, and observed that when we recycled the app pool, the requests were queued by IIS and eventually served, with no 503s. This behavior exactly matches our .Net Framework services, both with Overlapped Recycle enabled or disabled (disabling Overlapped Recycle simply causes a longer queue time since the old process has to fully exit before the new one can start up). We didn't touch the shutdown delay setting, and IIS reliably queued the incoming requests without any 503s, so that default seems sufficient. 

I don't think we would roll a DLL tweak like this out to production to really test the stability like you asked for, but I hope the other devs in this thread and the previous thread give it a try. I'm also hoping this fix can land as a patch for .Net 6 in addition to 8, so we can take advantage of it without having to rush our upgrades.

**Comment by divil5000** [comment]
That's really encouraging to hear. Like you though I cannot afford to deploy this into production, unless there's a straightforward (and easily reversible) way of doing so. Brennan?

**Comment by MV10** [comment]
@ckobelski thanks for posting that. If you don't mind, could you elaborate on your testing? Did you watch perf counters and saw requests queued? Was the test designed to simulate an unhealthy old pool, for example, hung and no longer processing requests? Could you test other recycle scenarios like WAS automated recycles due to exceeding the private bytes threshold?

Unfortunately my org couldn't touch this (even for testing) unless it was officially supported. And per-app settings in `web.config` is a deal-breaker given that we support tens of thousands of apps, many of which are quite large and on busy shared servers, and could easily take more than 1000ms to spin up -- but it's nice to know a solution may be possible.

I assume a true fix would require IIS changes, I'm guessing this behavior is tied to managed-pool support within IIS itself, and perhaps this is the real reason MS has been unwilling to fix this properly.

**Comment by adityamandaleeka** [comment]
Thanks @ckobelski for trying this out!

@MV10 To clarify, we _are_ hoping to get this change into the product since we've heard lots of reports of this being a problem. We'd love to hear more people's experience with the DLL @BrennanConroy shared to help us gain confidence in the approach.

Of course, we understand that not everyone is comfortable trying out an experimental fix.

**Comment by divil5000** [comment]
More people will be willing/able to try this fix if you make it crystal clear how we can easily do it, in a reversible way, so we are confident we aren't going to break anything. This problem really only manifests itself with servers under reasonable load.

**Comment by BrennanConroy** [comment]
> if you make it crystal clear how we can easily do it, in a reversible way

As posted originally:

> To use the dll below (assuming x64), replace the dll in C:\Program Files\IIS\Asp.Net Core Module\V2\ (keep the original around so you can switch back to the supported dll later).

This is how you try the fix and reverse it. Just swap the dlls. For example, I rename the original one to aspnetcorev2-orig.dll so it's in the original location just with a different name.
<img width="413" alt="image" src="https://github.com/dotnet/aspnetcore/assets/7574801/45142f8f-3118-49f8-a140-011bfddb1df1">
If you lose the original dll, you should be able to get it back by repairing the hosting bundle installation or [downloading the hosting bundle](https://dotnet.microsoft.com/en-us/download/dotnet/8.0) and reinstalling it (probably want to delete the aspnetcorev2.dll first).

**Comment by divil5000** [comment]
Ok, thank you. I assume I will also have to stop all asp.net websites while doing this, then simply restart them as soon as the dll is replaced, and that I need to be using asp.net 8.

**Comment by BrennanConroy** [comment]
ASP.NET Core 8 is not needed. Any (supported) version should work.

**Comment by ckobelski** [comment]
@MV10 
> @ckobelski thanks for posting that. If you don't mind, could you elaborate on your testing? Did you watch perf counters and saw requests queued? Was the test designed to simulate an unhealthy old pool, for example, hung and no longer processing requests? Could you test other recycle scenarios like WAS automated recycles due to exceeding the private bytes threshold?

My testing was quick and dirty and not a robust scenario, but enough to convince me that the fix was on the right track. I set up several powershells with `while($true){$ProgressPreference='SilentlyContinue';Invoke-WebRequest -UseBasicParsing https://my-service/health-endpoint}` to simulate a constant load of a few users, recycled the application pool for my service in IIS with just right click -> Recycle, and checked the IIS logs for logged status codes and time taken. I did that test a couple of times for each of: a .Net Framework service in IIS, a .Net 6 service in IIS with the existing ANCM, and a .Net 6 service in IIS with the updated DLL, and compared the results across the three setups. 

With the existing ANCM, recycle caused IIS to return 503s for a brief period of time after the recycle, then queue requests waiting for the new process to finish come up. With .net framework and with the new DLL, requests coming in after the recycle would just queue for the new process, no 503s. I could tell they were queued by IIS waiting for the new process because they did eventually get served, and the time taken logged by IIS reflected the wait. Since the new behavior matches .net framework's behavior, to me this means the regression in functionality has been addressed.

I didn't test what happens if the application hangs or crashes, because the application would already be in a bad state, so another second or two of 503s after an extended period of 502s and 500s doesn't make a big difference. Memory-based and periodic recycling is just the system issuing a recycle instead of a user, so it shouldn't be any different than manually commanding it.

**Comment by cun-dp** [comment]
Thanks @BrennanConroy for this holiday present 😊

I re-ran my tests with your experimental aspnetcorev2.dll with promising results, and some strange new issues I observed during testing. 

### Test setup:
* My application under test is an almost unchanged dotnet 6 aspnet webapi template, with a controller action that just responds with a static string to get requests. 
* All tests were done using in-process hosting unless noted otherwise. 
* The critical addition is a “Application Stopped” event handler:
`app.Lifetime.ApplicationStopped.Register(() => { Thread.Sleep(TimeSpan.FromMilliseconds(1000)); });`
* I did not use the new “shutdownDelay” setting unless noted otherwise.
* The tests were run on a VM with 4 CPU cores.
* I used an endless loop with Invoke-WebRequest in PowerShell to simulate load on the application.

### The good:
I have **not** been able to cause 503 responses with this dll, no matter how I recycled, no matter if in-proc or out-of-proc hosting. Before, I got ~60 503s with the stable dll, using the same test setup.

### The issue:
The old apppool instance did not shutdown correctly most (all?) of the time. It either crashed hard, or hangs with high CPU load. Which case happened seemed to depend on the sleep timespan in the application stopped event handler with high reproducibility (though <100%).

Neither case did trigger 503s on its own.

### Crash case:
The old apppool crashes about 20s after the recycle when I used 10000ms for the sleep timespan. It did not use excessive CPU time during this time. This could trigger RFP if done frequently, and therefor indirectly cause 503s.

**System Event Logs:**
Source: WAS
`A process serving application pool 'test-<PII censored>-site-aspnetdemo' terminated unexpectedly. The process id was '524'. The process exit code was '0xc0000005'.`

**Application Event Logs:**	
Source: IIS AspNetCore Module V2	
`Failed to gracefully shutdown application 'MACHINE/WEBROOT/APPHOST/TEST-<PII censored>-SITE/ASPNETDEMO'`

Source: .NET Runtime	
```
Application: w3wp.exe
CoreCLR Version: 6.0.2523.51912
.NET Version: 6.0.25
Description: The process was terminated due to an unhandled exception.
Exception Info: exception code c0000005, exception address 00007FF920EF6A77
Stack:
```

Source: Application Error	
```
Faulting application name: w3wp.exe, version: 10.0.14393.0, time stamp: 0x57899b8a
Faulting module name: iiscore.dll, version: 10.0.14393.6343, time stamp: 0x650281fb
Exception code: 0xc0000005
Fault offset: 0x0000000000026a77
Faulting process id: 0x20c
Faulting application start time: 0x01da3f1c913a4a57
Faulting application path: c:\windows\system32\inetsrv\w3wp.exe
Faulting module path: C:\Windows\system32\inetsrv\iiscore.dll
Report Id: 81c045ca-3ef0-4bff-8b27-b713e58179a1
Faulting package full name: 
Faulting package-relative application ID:
```

### CPU burn case:
When I used 1000ms (or less) for the sleep timespan, the old apppool instance hangs while using 100% of one cpu core, until the the apppool shutdown time limit (90s in my case) kills it. This seemed to only happen if the apppool had received at least one request before recycling.

![image](https://github.com/dotnet/aspnetcore/assets/75260314/7e7d84a6-4ca3-4ecb-ada0-ab28123e5d9c)

 
Recycling 4 times in short order pegged all 4 cores at 100%, making the VM completely unresponsive until the first pool got killed. Requests where served without interruption though (I did not measure throughput).

When I set the “shutdownDelay” setting to 20s, the CPU burn started 20s after the recycle, instead of almost instantly (1s?). 
The application event log said “Application 'MACHINE/WEBROOT/APPHOST/TEST-<PII censored>-SITE/ASPNETDEMO' has shutdown.” 20s after the recycle, although the old apppool process was still alive, burning CPU cycles.
<br/><br/>
I hope this makes sense?

**Comment by BrennanConroy** [comment]
@cun-dp thanks for trying this out!

I've been unable to reproduce the crashes and CPU burning. The only thing I get when sleeping for 10 seconds during shutdown is the `"Failed to gracefully shutdown"` log. Could you provide information about the OS you're using? I'll also switch my test app to 6.0 with MVC (I'm using 9.0 with plain middleware) to see if that changes anything, although I would be surprised.

If you could get a crash dump and share it, that would also be helpful.

**Comment by cun-dp** [comment]
I made 2 dumps. These are multipart zip archives created with 7zip. I had to add ".TXT" as a fake filename extension in order for github to allow uploading.

* CPU_burn_w3wp.exe_240108_104804: Process dump during CPU burn (via `procdump64.exe -ma` )
* Crash_w3wp.exe_240108_104949: Crash dump (via `procdump64.exe -e -ma` )

[Crash_w3wp.exe_240108_104949.zip.001.TXT](https://github.com/dotnet/aspnetcore/files/13858825/Crash_w3wp.exe_240108_104949.zip.001.TXT)
[Crash_w3wp.exe_240108_104949.zip.002.TXT](https://github.com/dotnet/aspnetcore/files/13858828/Crash_w3wp.exe_240108_104949.zip.002.TXT)
[CPU_burn_w3wp.exe_240108_104804.zip.002.TXT](https://github.com/dotnet/aspnetcore/files/13858836/CPU_burn_w3wp.exe_240108_104804.zip.002.TXT)
[CPU_burn_w3wp.exe_240108_104804.zip.003.TXT](https://github.com/dotnet/aspnetcore/files/13858840/CPU_burn_w3wp.exe_240108_104804.zip.003.TXT)
[CPU_burn_w3wp.exe_240108_104804.zip.001.TXT](https://github.com/dotnet/aspnetcore/files/13858841/CPU_burn_w3wp.exe_240108_104804.zip.001.TXT)

OS Infos (excerpt from `Get-ComputerInfo` output):
```
OsName                                     : Microsoft Windows Server 2016 Datacenter
OsType                                     : WINNT
OsOperatingSystemSKU                       : DatacenterServerEdition
OsVersion                                  : 10.0.14393
OsBuildNumber                              : 14393
OsCountryCode                              : 49
OsCurrentTimeZone                          : 60
OsLocaleID                                 : 0407
OsLocale                                   : de-DE
OsBuildType                                : Multiprocessor Free
OsCodeSet                                  : 1252
OsDataExecutionPreventionAvailable         : True
OsDataExecutionPrevention32BitApplications : True
OsDataExecutionPreventionDrivers           : True
OsDataExecutionPreventionSupportPolicy     : OptOut
OsDebug                                    : False
OsDistributed                              : False
OsEncryptionLevel                          : 256
OsForegroundApplicationBoost               : Maximum
OsTotalVisibleMemorySize                   : 8388016
OsFreePhysicalMemory                       : 4860672
OsTotalVirtualMemorySize                   : 9698736
OsFreeVirtualMemory                        : 6032180
OsInUseVirtualMemory                       : 3666556
OsSizeStoredInPagingFiles                  : 1310720
OsFreeSpaceInPagingFiles                   : 1310720
OsPagingFiles                              : {C:\pagefile.sys}
OsHardwareAbstractionLayer                 : 10.0.14393.5786
OsInstallDate                              : 11.03.2019 15:15:53
OsManufacturer                             : Microsoft Corporation
OsMaxNumberOfProcesses                     : 4294967295
OsMaxProcessMemorySize                     : 137438953344
OsMuiLanguages                             : {en-US}
OsNumberOfLicensedUsers                    : 0
OsNumberOfProcesses                        : 92
OsNumberOfUsers                            : 18
OsArchitecture                             : 64-bit
OsLanguage                                 : en-US
OsProductSuites                            : {TerminalServices, DatacenterEdition, TerminalServicesSingleSession}
OsPortableOperatingSystem                  : False
OsPrimary                                  : True
OsProductType                              : Server
OsServicePackMajorVersion                  : 0
OsServicePackMinorVersion                  : 0
OsStatus                                   : OK
OsSuites                                   : {TerminalServices, DatacenterEdition, TerminalServicesSingleSession}
OsServerLevel                              : FullServer
```

**Comment by cun-dp** [comment]
For completeness sake: Here is my test code. 

[WebApiDemo.zip](https://github.com/dotnet/aspnetcore/files/13863211/WebApiDemo.zip)


* Bootstrap/App Start/App Stopp delays can be configured via `appsettings.json` , see `Program.cs`.
* Is used the endpoint `~/ping` for my recent tests.

**Comment by BrennanConroy** [comment]
Ok, I reproed the crash and CPU burn on a Win Server 2016 machine and verified a fix. Below is a new build with the additional fix.

**Original instructions and details:**
https://github.com/dotnet/aspnetcore/issues/41340#issuecomment-1857250264

**New build:**
SHA256: `49CA86EDC21AAE799F836F7C348E090126F7A9CE2541E38E9AD6F4CD646FDA91`
[aspnetcorev2.zip](https://github.com/dotnet/aspnetcore/files/13983076/aspnetcorev2.zip)

**Comment by cun-dp** [comment]
Looks great! I have not been able to reproduce the issues from https://github.com/dotnet/aspnetcore/issues/41340#issuecomment-1878562958

Even when setting the ApplicationStopped event handler timeout to >1000ms, no 503s occured, only the expected event log messages `Failed to gracefully shutdown application`, which could be circumvented by increasing the new `shutdownDelay` setting.

Sidenote: 
Apparently, my VM got some patches in the meantime, updating the stable aspnetcorev2.dll from to `18.0.23305.0` to `18.0.23334.1`. This made the problem worse by a factor of 10,  i.e. 10 times the 503 responses per apppool recycle.

**Comment by adityamandaleeka** [comment]
@cun-dp That's great to hear, and thank you for the excellent detailed feedback and investigations!

**Comment by mtanksl** [comment]
@BrennanConroy Does this change take [Shutdown Timeout](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/host/web-host?view=aspnetcore-3.1#shutdown-timeout) configuration from ASP.NET Core Web Host into consideration? or should we not worry about it?

**Comment by BrennanConroy** [comment]
`shutdownDelay` affects Shutdown Timeout in the sense that it delays telling the server to shutdown which is when Shutdown Timeout will kick in. So it increases the process shutdown time, but not the app shutdown time.

`shutdownDelay` would affect the AppPool shutdown timeout though. In the sense that if `shutdownDelay` + server shutdown now takes longer than the AppPool shutdown timeout it'll be terminated.
e.g. setting `shutdownDelay` to 91 seconds would terminate the w3wp process and never even start shutting down the app.

![image](https://github.com/dotnet/aspnetcore/assets/7574801/d0d207db-425a-4773-8f73-55a0f656833c)

**Comment by mtanksl** [comment]
@BrennanConroy Perfect, got it. After testing the dll for a few hours in production, no 503 error appeared. It seems to be working perfectly. Thank you.

**Comment by paddlepaw** [comment]
The experimental dll works for us only if we extend the shutdownDelay.  Before that our Umbraco CMS(ver 9) site would still return 503s on recycle so the dll seems to be working. Thanks.

Now it only goes down when we deploy the site and I don't have a fix for that yet because it's not happening every time.  I tried installing Application Initialization and turning on preload but that seems to make it worse in that I have to recycle or shutdown/restart the app pool more times to bring it back online.  We are making use of app_offline.htm on deploy and I can see a recycle event occur in the event log when it detects app_offline.htm.  It'll show the app shutdown and restart successfully most of the time but sometimes I get an error that it failed to shutdown gracefully.  The only recourse is to shut down the app pool and then restart it and normally when this error occurs I have to shutdown/restart multiple times.  I'll keep working on it but just wanted to put my experience out there.

**Comment by randyshoopman** [comment]
@BrennanConroy  It's great to hear this really old issue is finally seeing a resolution.   Since work is being done on this is there anyway to improve the error:

```
Failed to gracefully shutdown application 'MACHINE/WEBROOT/APPHOST/XXXX'.
Process Id: 9112.
File Version: 17.0.23296.14. Description: IIS ASP.NET Core Module V2 Request Handler. Commit: 0a715692d8e2536c899faa0bb4f0cec2c2e33e36
```
Like adding something to indicate what request or ApplicationStopped callback is holding up the shutdown?  It would really go a long way to help address problematic code.  Request Route, stack trace, anything to know what code is blocking the shutdown.

Thanks!

**Comment by divil5000** [comment]
@randyshoopman what a brilliant idea. We've always suffered with that error, and despite wasting hours trying to figure out what's happening, have not been successful in finding the source. We just grudgingly accept it.

**Comment by paddlepaw** [comment]
An update to my last post.  So I thought the DLL fix was working for us but it turns out that manual recycles no longer cause crashes but automatic recycles do.  I left automatic recycling on with the default values for the last 3 days. Checking the IIS logs, the site went down 3 days ago until this morning when the normal daily 4 a.m. recycle brought it back online.  We have another vanilla .net core site that is not crashing at all so this leads me to believe it's either something that we have implemented ourselves on this particular site or Umbraco itself.

**Comment by Phlow2001** [comment]
Still waiting and still affecting production. What's the status of this? What do we have to do to get some priority and resources on it? Is it on the team's radar at all? Is there a plan to gain confidence in the approach in the modified v2 module and then to get it published officially in a hosting bundle update?

**Comment by adityamandaleeka** [comment]
Yes, as mentioned above, it's not only on our radar, but we intend to release this fix in .NET 9.

We're also trying to figure out a way to include the fix in a .NET 8 servicing release.

**Comment by cun-dp** [comment]
> Yes, as mentioned above, it's not only on our radar, but we intend to release this fix in .NET 9.
> 
> We're also trying to figure out a way to include the fix in a .NET 8 servicing release.

Thanks, that sounds promising. We would really appreciate a fix for .NET8, as we are only able to run LTS releases.

**Comment by MV10** [comment]
We're in the same situation, only LTS releases are permitted. What would happen if it is only released for .NET 9 and later, but a .NET 8 application was running behind the new version? If it isn't available for .NET 8, this is actually pretty likely to happen for us once .NET 10 is approved for internal use. I suppose the alternate question is, can .NET 9 and later run under the old (now current) ANCM until everything is updated to a version supported by the one with recycle improvements?

(Hopefully it'll make the cut and my questions will be completely irrelevant!)

**Comment by adityamandaleeka** [comment]
Good question @MV10. The particular component that the fix is in is actually shared across installs, so our approach to servicing it is to ensure that it is backwards compatible with older supported versions.

This comes with ups and downs obviously... on the bright side, fixes that we make can apply down-level easily. On the other hand, we have to be _super careful_ to ensure that we don't break existing apps. We don't want anyone's existing app to break without them doing anything just because they installed a newer hosting bundle on their server.

This is why we're being careful and ensuring we have enough bake time and testing (and considering things like opt-in switches for down-level) before getting this out there. We have heard loud and clear that people want this fix and we can't wait to solve this problem for everyone.

**Comment by bluntspoon** [comment]
Thanks for starting the PR for this issue @BrennanConroy. Keen to know which versions of DotNet will get this fix once merged. Seeing updates to dotnet6+ would be awesome.

**Comment by BrennanConroy** [comment]
This will be opt-out in 9.0 and we're backporting as opt-in for 8.0. Since the change is in the IIS module which is installed globally on the machine and the module is compatible with all supported versions of ASP.NET Core you can use the 8.0 or 9.0 module with 6.0 and still get the fix.

When installing the hosting bundle (which is what installs the IIS module) you can choose to ignore the runtime that comes with the installer if you want to continue using 6.0 but get the newer module. See https://learn.microsoft.com/aspnet/core/host-and-deploy/iis/hosting-bundle?view=aspnetcore-8.0#options for details.

**Comment by RomBrz** [comment]
Great to see this problem being solved! As the Hosting Bundle packages, some time ago i tried that "choice to ignore the runtime" but for some reason "Install Hosting Bundle with runtimes and after that uninstall both runtimes" gave me a different setup than "Install Hosting Bundle only without runtime". Even the size of the Hosting Bundle was different. Seems to persist this "bug":
(After Install with OPT_NO_RUNTIME=1)
![image](https://github.com/dotnet/aspnetcore/assets/24864710/811cdc6f-17a7-455c-85a5-3625ec031a35)

(After Full Install)
![image](https://github.com/dotnet/aspnetcore/assets/24864710/4ede13d0-c0c6-4828-a1c2-59cbdb1d48d2)

On our servers all the DotNet core apps only works with the 145MB Server Hosting (so i have to do a full install and after that remove the runtimes).

Considering that you are working with ANCM, some time ago i posted about this problem involving IIS, DotNet Core and Windows Update (https://github.com/dotnet/aspnetcore/issues/41377). The problem persists as for some reason i need to put IIS in manual mode, reboot server, than do the Windows Update full install (Windows security patches, Dotnet Framework security patches and DotNet Core patches).

**Comment by MV10** [comment]
A little bit off-topic, but until we have this fix deployed, our monitoring is being configured to ignore _Failed to gracefully shut down application XYZ_ error messages (which correlate to the recycle 503s, of which there are many since corporate policy still requires early-AM staggered-schedule recycles of all pools).

Is there a reference somewhere of the various warnings and errors ANCM can emit?

**Comment by simon-biber** [comment]
Did this fix get into .NET 8.0.5 hosting bundle?

**Comment by MV10** [comment]
The way I read it, it'll be released with .NET 9 in November, and backported to .NET 8.x.x at that time...

**Comment by cun-dp** [comment]
According to milestone tags the fix ist slated for the next release 8.0.6: https://github.com/dotnet/aspnetcore/issues?q=milestone%3A8.0.6%20is%3Aclosed%20label%3Aservicing-approved

🤞

**Comment by adityamandaleeka** [comment]
@cun-dp is right, it'll ship in the 8.0.6 release in June.

**Comment by BrennanConroy** [comment]
This fix is out in 9.0-preview4 now. And will be available in 8.0.6 in June-2024.

I'm going to close and lock this issue so that this comment, which will include detailed instructions below, is the last one. For those of you who have been posting semi-unrelated issues, file a new issue this isn't the place for it.

### Instructions

When installing the hosting bundle (which is what installs the global IIS module) you can choose to ignore the runtime that comes with the installer if you want to continue using an older runtime but get the newer modules improvements. See https://learn.microsoft.com/aspnet/core/host-and-deploy/iis/hosting-bundle?view=aspnetcore-8.0#options for details. Otherwise, install the hosting bundle as normal to get the IIS module installed.

There are 2 ways to modify the behavior of this change.
1. Modify the `shutdownDelay` option in your web.config `handlerSettings`.
```xml
<aspNetCore processPath="dotnet" arguments="myapp.dll" stdoutLogEnabled="false" stdoutLogFile=".logsstdout">
  <handlerSettings>
    <!--
    Milliseconds to delay shutdown of the old app app instance while the new instance starts.
    Note: This doesn't delay the handling of incoming requests.
    -->
    <handlerSetting name="shutdownDelay" value="5000" />
  </handlerSettings>
</aspNetCore>
```
2. Set the `ANCM_shutdownDelay` environment variable, also in milliseconds.

If you set the value to 0, then it goes back to the old behavior.

For `9.0 >=`, the fix is on by default and can be disabled by setting the config to `0`.

For `<= 8.0`, the fix is off by default and can be enabled by setting the config to a value `> 0`.

The default is 1000 milliseconds (1 second). For busy/slow machines you may want to increase this value.
